api/staged_products_property_collection_helpers.go

```diff
@@ -0,0 +1,142 @@
+package api
+
+import (
+	"fmt"
+	"regexp"
+	"sort"
+)
+
+type updatedPropertyCollectionItem struct {
+	Data map[string]interface{}
+}
+type updatedPropertyCollection []updatedPropertyCollectionItem
+
+func (item updatedPropertyCollectionItem) getFieldValue(fieldName string) string {
+	if value, ok := item.Data[fieldName].(string); ok {
+		return value
+	}
+	return ""
+}
+
+func (item updatedPropertyCollectionItem) setFieldValue(fieldName string, value string) {
+	item.Data[fieldName] = value
+}
+
+func parseUpdatedPropertyCollection(updatedProperty interface{}) (updatedPropertyCollection, error) {
+	var collection updatedPropertyCollection
+
+	if updatedPropertyAsMap, ok := updatedProperty.(map[string]interface{}); ok {
+		rawItems := updatedPropertyAsMap["value"]
+		if rawItemSlice, ok := rawItems.([]interface{}); ok {
+			for _, item := range rawItemSlice {
+				if itemMap, ok := item.(map[string]interface{}); ok {
+					collection = append(collection, updatedPropertyCollectionItem{Data: itemMap})
+				} else {
+					return nil, fmt.Errorf("parseUpdatedPropertyCollection: failed to convert %v to map[string]interface{}", item)
+				}
+			}
+		} else {
+			return nil, fmt.Errorf("parseUpdatedPropertyCollection: failed to convert %v to []interface{}", rawItems)
+		}
+	} else {
+		return nil, fmt.Errorf("parseUpdatedPropertyCollection: failed to convert %v to map[string]interface{}", updatedProperty)
+	}
+	return collection, nil
+}
+
+type responsePropertyCollection []responsePropertyCollectionItem
+type responsePropertyCollectionItem struct {
+	Data map[interface{}]interface{}
+}
+
+func parseResponsePropertyCollection(rawItems interface{}) (responsePropertyCollection, error) {
+	var collection responsePropertyCollection
+	if rawItemSlice, ok := rawItems.([]interface{}); ok {
+		for _, item := range rawItemSlice {
+			if itemMap, ok := item.(map[interface{}]interface{}); ok {
+				collection = append(collection, responsePropertyCollectionItem{Data: itemMap})
+			} else {
+				return nil, fmt.Errorf("parseResponsePropertyCollection: failed to convert %v to map[interface{}]interface{}", item)
+			}
+		}
+	} else {
+		return nil, fmt.Errorf("parseResponsePropertyCollection: failed to convert %v to []interface{}", rawItems)
+	}
+	return collection, nil
+}
+
+func (item responsePropertyCollectionItem) getFieldValue(fieldName string) string {
+	if valueObj, ok := item.Data[fieldName].(map[interface{}]interface{}); ok {
+		if value, ok := valueObj["value"].(string); ok {
+			return value
+		}
+	}
+	return ""
+}
+
+//Find and return first string matching regex pattern
+//returns matched <string, true> if found, or <"",false> if not
+func findFirst(strings []string, fieldRegex string) (string, bool) {
+	r, err := regexp.Compile(fieldRegex)
+	if err != nil {
+		panic(fmt.Sprintf("failed to compile regex: %v because:\n%v", fieldRegex, err))
```

Comment:
we shouldn't panic in the code (even if the code shouldn't fail). We're going to have this function return an error, and exit based on the error instead

Reply:
I felt dirty doing this; but went ahead since I felt this was an error that could only occur at "compile time" (by the programmer using the function passing in "bad" regex)

The refactoring of `findLogicalKeyField()` proposed below sidesteps this issue; since the regexes in question are in the same function.  (Interestingly; I believe that `MustCompile()` will also `panic` if the regex is invalid - https://golang.org/src/regexp/regexp.go?s=10768:10804#L298)